### PR TITLE
[#9230][feat] Slimmed down implementation of nemotron H

### PR DESCRIPTION
**Pull request by 2ez4bz**
## Summary by CodeRabbit

* **New Features**
  * Added public API to register custom model implementations via `register_custom_model_cls()` method for flexible model extensibility.
  * Introduced support for NemotronH model with Mamba-inspired architecture, including specialized mixer, attention, and normalization components.

* **Tests**
  * Added test coverage for custom model registration and building workflows.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### Why?

The reference nemotron H code on HuggingFace is out of date, and therefore bugged, and has several untested code paths. This makes an already hairy patching system even hairier.

The proposal is to do away with those patches, and replace the original implementation with one that is heavily slimmed down.

### What?

This PR sets the basis for an alternative path with such an slimmed down implementation that:

* fixes bugs in the current HF implementation
* adds no new dependencies to TensorRT-LLM
* does away with unnecessary features for TensorRT-LLM / AutoDeploy:
  - no training related code (dropout, gradient checkpointing, etc.)
  - no caching business (we want to replace it with our own anyway)
  - no attention masking where possible
* reuses existing AD custom ops for mamba SSM update / causal conv1d

In order for the above to be usable in the AD apparatus, some extensions to its `LlmArgs` and `AutoModelForCausalLMFactory` are made.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py [hunk 1]
```diff
@@ -0,0 +1,826 @@
...
+    """
+    Base class for causal language model (or autoregressive) outputs.
+
+    Args:
+        logits (`torch.FloatTensor` of shape `(batch_size, sequence_length, config.vocab_size)`):
+            Prediction scores of the language modeling head (scores for each vocabulary token before SoftMax).
+    """
+
+    logits: Optional[torch.FloatTensor] = None
+    attentions: Optional[Tuple[torch.FloatTensor]] = None
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py]: Ditto.

File: tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py [hunk 2]
```diff
@@ -0,0 +1,826 @@
...
+    def __init__(self, config):
+        super().__init__(config)
+        self.backbone = NemotronHModel(config)
+        self.vocab_size = config.vocab_size
+        self.lm_head = nn.Linear(config.hidden_size, config.vocab_size, bias=False)
+
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    def get_input_embeddings(self):
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py]: Check if any of these can be removed.

File: tensorrt_llm/_torch/auto_deploy/models/hf.py
Comment (lucaslie) [inline tensorrt_llm/_torch/auto_deploy/models/hf.py]: I would keep this part simpler:

1. @register decorator for custom model code what should be re-routed (e.g. `AutoModelForCausalLM` or another auto class) and when it should be re-routed (e.g. based on class name of original implementation, e.g.,  `NemotronHForCausalLM`) 
```
@register_custom_modeling_class("AutoModelForCausalLM")
class NemotronHForCausalLM:
    ...
```
2. In `hf.py` we simply re-route to the custom modeling class's `._from_config` if conditions are satisfied

File: tensorrt_llm/_torch/auto_deploy/llm_args.py
Comment (lucaslie) [inline tensorrt_llm/_torch/auto_deploy/llm_args.py]: I don't think we should make this configurable. Either we maintain a custom model implementation or we don't. I don't think it's good to have discrepancy in the behavior or give the user a choice which may require to unnecessarily support both paths (directly from HF checkpoint or from custom modeling code)

File: tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py [hunk 3]
```diff
@@ -0,0 +1,826 @@
...
+        )
+
+
+# Forked from:
+# https://github.com/state-spaces/mamba/blob/6b32be06d026e170b3fdaf3ae6282c5a6ff57b06/mamba_ssm/ops/triton/layernorm_gated.py
+# NOTES:
+# 1. At time of writing (09/25/2025), the nano nemotron v2 modeling code expects `mamba_ssm`
+#    to be installed so as to be able to make use of its grouped gated RMS norm operation.
+#    We therefore replace it with one that uses einops + pytorch.
+def _rms_norm_ref(
```
Comment (lucaslie) [inline tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py]: probably easier to just call the custom op directly instead of maintaining a ref version here

File: tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py [hunk 4]
```diff
@@ -0,0 +1,826 @@
...
+        attn_output = torch.nn.functional.scaled_dot_product_attention(
+            query_states,
+            key_states,
+            value_states,
+            attn_mask=None,
+            # Hardcoding to 0.0 since we should always be in eval mode.
+            dropout_p=0.0,
+            is_causal=True,
+        )
+        attn_output = attn_output.transpose(1, 2).contiguous()
```
Comment (lucaslie) [inline tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py]: we can just do that instead
```suggestion
        query_states = query_states.view(bsz, q_len, self.num_heads, self.head_dim)
        key_states = key_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim)
        value_states = value_states.view(
            bsz, q_len, self.num_key_value_heads, self.head_dim
        )


        attn_output = torch.ops.auto_deploy.torch_attention(
            query_states,
            key_states,
            value_states,
            attn_mask=None,
            # Hardcoding to 0.0 since we should always be in eval mode.
            dropout_p=0.0,
            is_causal=True,
            layout="bsnd",
        )
```

File: tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py [hunk 5]
```diff
@@ -0,0 +1,826 @@
...
+                "input_ids": input_ids.contiguous()
+            }  # `contiguous()` needed for compilation use cases
+
+        model_inputs.update(
+            {
+                "position_ids": position_ids,
+                "logits_to_keep": self.config.num_logits_to_keep,
+            }
+        )
+        return model_inputs
```
Comment (lucaslie) [inline tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py]: remove

File: tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py (hunk 2 above)
Comment (2ez4bz) [reply tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py]: Our of curiosity, why?

Comment (lucaslie) [reply tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py]: Two particular places where those utilities may be used that I have found so far:
1. Different parts of the model may use the embedding layer to go from input_ids --> input_embeds. For example, in VLMs this is dependent on whether images are present.
2. tieing model weights between the lm head and the embedding also rely on those utilites afaik.

I found those to be nasty to hunt down - so not worth cleaning up imo and trying to figure when which of those is used

File: tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py (hunk 5 above)
Comment (lucaslie) [reply tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py]: yes, not needed

File: tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py [hunk 6]
```diff
@@ -0,0 +1,826 @@
...
+import torch.utils.checkpoint
+from einops import rearrange
+from torch import nn
+from transformers.activations import ACT2FN
+from transformers.configuration_utils import PretrainedConfig
+from transformers.generation import GenerationMixin
+from transformers.modeling_utils import PreTrainedModel
+from transformers.utils import ModelOutput, logging
+
+logger = logging.get_logger(__name__)
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py]: Done.

File: tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py [hunk 7]
```diff
@@ -0,0 +1,826 @@
...
+class NemotronHRMSNorm(nn.Module):
+    def __init__(self, hidden_size, eps=1e-6):
+        """
+        NemotronHRMSNorm is equivalent to T5LayerNorm and LlamaRMSNorm
+        """
+        super().__init__()
+        self.weight = nn.Parameter(torch.ones(hidden_size))
+        self.variance_epsilon = eps
+
+    def forward(self, hidden_states):
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py]: Yeah this one's not gated.

File: tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py (hunk 3 above)
Comment (lucaslie) [reply tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py]: nevermind this comment. Let's keep as is. I see one issue doing this:

rms norm pattern detection and backend selection is in one go. So we have no good way of changing the kernel selection if we don't invoke the pattern matcher

File: tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py [hunk 8]
```diff
@@ -0,0 +1,826 @@
...
+    """
+    Class for the NemotronH model outputs.
+
+    Args:
+        last_hidden_state (`torch.FloatTensor` of shape `(batch_size, sequence_length, hidden_size)`):
+            Sequence of hidden-states at the output of the last layer of the model.
+    """
+
+    last_hidden_state: Optional[torch.FloatTensor] = None
+    attentions: Optional[Tuple[torch.FloatTensor]] = None
```
[the remark this answers is not in the document]
Comment (2ez4bz) [reply tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py]: Got some errors from transformers internals if I removed the class entirely, so sticking with them for now.

However, `attentions` can be (and have been) safely removed.

File: tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py (hunk 2 above)
Comment (2ez4bz) [reply tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py]: Thanks for the context. Keeping.

File: tensorrt_llm/_torch/auto_deploy/models/hf.py
Comment (2ez4bz) [reply tensorrt_llm/_torch/auto_deploy/models/hf.py]: Updated as we discussed on slack.

File: tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py [hunk 9]
```diff
@@ -0,0 +1,543 @@
...
+            [
+                self.intermediate_size,
+                self.n_groups * self.ssm_state_size,
+                self.n_groups * self.ssm_state_size,
+            ],
+            dim=-1,
+        )
+
+        # 3. SSM transformation
+        A = -torch.exp(self.A_log.float())  # [num_heads]
```
Comment (suyoggupta) [inline tensorrt_llm/_torch/auto_deploy/models/modeling_nemotron_h.py]: is there a way to avoid doing this in the forward ?

**Review by govind-ramnarayan — APPROVED** [review]
LGTM, some small nits

File: tests/unittest/_torch/auto_deploy/unit/singlegpu/models/test_hf.py [hunk 10]
```diff
@@ -123,3 +132,54 @@ def test_recursive_update_config(mock_factory):
...
+    )
+
+    assert AutoModelForCausalLMFactory._custom_model_mapping[model_type] == custom_model_cls
+
+
+class MyError(Exception):
+    pass
+
+
+def test_build_model_uses_custom_model_cls_from_config(mock_factory):
```
[the remark this answers is not in the document]
Comment (govind-ramnarayan) [reply tests/unittest/_torch/auto_deploy/unit/singlegpu/models/test_hf.py]: In retrospect I think it's probably better as-is
